File: runme/mcgen/local/datan/datsrc/minenc.c
#include "datsrc.h"

void LIBFUNC minenc_(double P_T *xa, double P_T *xb, double P_T *xc,
             double P_T *ya, double P_T *yb, double P_T *yc,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), integer P_T *nstep,
             double const P_T *x0, double const P_T *xdir,
             integer const P_T *n)
{
    double xend;
    integer istep, nst;
    double xm, ym, buf;

    /* Parameter adjustments */
    --xdir;
    --x0;

    /* Function Body */
    *ya = minfnd_(xa, &x0[1], &xdir[1], n, userfn);
    *yb = minfnd_(xb, &x0[1], &xdir[1], n, userfn);
    nst = *nstep;
    if (nst <= 0) {
        nst = 1000;
    }
    if (*yb > *ya) {
/* exchange A and B */
        buf = *xa;
        *xa = *xb;
        *xb = buf;
        buf = *ya;
        *ya = *yb;
        *yb = buf;
    }
    *xc = *xb + (*xb - *xa) * 1.618034;
    *yc = minfnd_(xc, &x0[1], &xdir[1], n, userfn);
    istep = 0;
L10:
    if (*yb >= *yc && istep < nst) {
        ++istep;
/* step was still downwards */
        xend = *xb + (*xc - *xb) * 10.;
        minprb_(&xm, xa, ya, xb, yb, xc, yc);
        ym = minfnd_(&xm, &x0[1], &xdir[1], n, userfn);
        if ((xm - *xb) * (*xc - xm) > 0.) {
/* XM is between XB and XC */
            if (ym < *yc) {
/* minimum is between XB and XC */
                *xa = *xb;
                *ya = *yb;
                *xb = xm;
                *yb = ym;
                goto L10;
            } else if (ym > *yb) {
/* minimum is between XA and XM */
                *xc = xm;
                *yc = ym;
                goto L10;
            }
/* there was no minimum, go on */
            xm = *xc + (*xc - *xb) * 1.618034;
        } else if ((*xc - xm) * (xm - xend) > 0.) {
/* XM is between XC and XEND */
            if (ym < *yc) {
                *xb = *xc;
                *xc = xm;
                xm = *xc + (*xc - *xb) * 1.618034;
                *yb = *yc;
                *yc = ym;
            }
        } else if ((xm - xend) * (xend - *xc) >= 0.) {
/* XM is beyond XEND */
            xm = xend;
        } else {
            xm = *xc + (*xc - *xb) * 1.618034;
        }
        *xa = *xb;
        *xb = *xc;
        *xc = xm;
        *ya = *yb;
        *yb = *yc;
        *yc = minfnd_(xc, &x0[1], &xdir[1], n, userfn);
        goto L10;
    }
    if (istep == nst) {
        *nstep = -1;
    } else {
        *nstep = istep;
    }
    return;
} /* minenc_ */

File: runme/mcgen/local/datan/datsrc/minenc.f
      SUBROUTINE MINENC(XA,XB,XC,YA,YB,YC,USERFN,NSTEP,X0,XDIR,N)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(ZERO=0.D0,GSPAR=1.618034D0,FACMAG=10.D0,
     +MAXSTP=1000)
      DIMENSION X0(N),XDIR(N)
      DOUBLE PRECISION MINFND
      EXTERNAL USERFN
      YA=MINFND(XA,X0,XDIR,N,USERFN)
      YB=MINFND(XB,X0,XDIR,N,USERFN)
      NST=NSTEP
      IF (NST.LE.0) NST=MAXSTP
      IF (YB.GT.YA) THEN
C exchange A and B
        BUF=XA
        XA=XB
        XB=BUF
        BUF=YA
        YA=YB
        YB=BUF
      END IF
      XC=XB+GSPAR*(XB-XA)
      YC=MINFND(XC,X0,XDIR,N,USERFN)
      ISTEP=0
   10 IF (YB.GE.YC .AND. ISTEP.LT.NST) THEN
        ISTEP=ISTEP+1
C step was still downwards
        XEND=XB+FACMAG*(XC-XB)
        CALL MINPRB(XM,XA,YA,XB,YB,XC,YC)
        YM=MINFND(XM,X0,XDIR,N,USERFN)
        IF ((XM-XB)*(XC-XM) .GT. ZERO) THEN
C XM is between XB and XC
          IF (YM.LT.YC) THEN
C minimum is between XB and XC
            XA=XB
            YA=YB
            XB=XM
            YB=YM
            GO TO 10
          ELSE IF (YM.GT.YB) THEN
C minimum is between XA and XM
            XC=XM
            YC=YM
            GO TO 10
          END IF
C there was no minimum, go on
          XM=XC+GSPAR*(XC-XB)
        ELSE IF ((XC-XM)*(XM-XEND) .GT. ZERO) THEN
C XM is between XC and XEND
          IF (YM.LT.YC) THEN
            XB=XC
            XC=XM
            XM=XC+GSPAR*(XC-XB)
            YB=YC
            YC=YM
          END IF
        ELSE IF ((XM-XEND)*(XEND-XC) .GE. ZERO) THEN
C XM is beyond XEND
          XM=XEND
        ELSE
          XM=XC+GSPAR*(XC-XB)
        END IF
        XA=XB
        XB=XC
        XC=XM
        YA=YB
        YB=YC
        YC=MINFND(XC,X0,XDIR,N,USERFN)
        GO TO 10
      END IF
      IF(ISTEP.EQ.NST) THEN
        NSTEP=-1
      ELSE
        NSTEP=ISTEP
      END IF
      END
